smart-hadoop-support/smart-hadoop/src/main/java/org/smartdata/hdfs/scheduler/SmallFileScheduler.java

```diff
@@ -0,0 +1,299 @@
+  public static final Logger LOG = LoggerFactory.getLogger(SmallFileScheduler.class);
+
+  public SmallFileScheduler(SmartContext context, MetaStore metaStore)
+      throws IOException {
+    super(context, metaStore);
+    this.metaStore = metaStore;
+    this.nnUri = HadoopUtil.getNameNodeUri(getContext().getConf());
+  }
+
+  @Override
+  public void init() throws IOException {
+    this.containerFilesLock = new ConcurrentHashMap<>();
+    this.containerFileMap = new ConcurrentHashMap<>();
+    this.fileContainerInfoMap = new ConcurrentHashMap<>();
+    this.smallFilesLock = new ArrayList<>();
+    this.smallFilesMap = new ConcurrentHashMap<>();
+    this.dfsClient = HadoopUtil.getDFSClient(nnUri, getContext().getConf());
+  }
+
+  @Override
+  public List<String> getSupportedActions() {
+    return ACTIONS;
+  }
+
+  private class ContainerFileExistInfo {
+    private String containerFilePath;
+    private boolean isExist;
+
+    private ContainerFileExistInfo(String containerFilePath, boolean isExist) {
+      this.containerFilePath = containerFilePath;
+      this.isExist = isExist;
+    }
+  }
+
+  @Override
+  public boolean onSubmit(ActionInfo actionInfo) {
+    if (ACTIONS.get(1).equals(actionInfo.getActionName())) {
+      long actionId = actionInfo.getActionId();
+
+      // Check if container file is null
+      String containerFilePath = actionInfo.getArgs().get(
+          SmallFileCompactAction.CONTAINER_FILE);
+      if (containerFilePath == null || containerFilePath.isEmpty()) {
+        return false;
+      }
+
+      // Check if small file list is null or empty
+      String smallFiles = actionInfo.getArgs().get(HdfsAction.FILE_PATH);
+      if (smallFiles == null || smallFiles.isEmpty()) {
+        return false;
+      }
+
+      // Get valid small file list according to the small file lock map
+      ArrayList<String> smallFileList = new Gson().fromJson(
+          smallFiles, new ArrayList<String>().getClass());
+      Iterator<String> iterator = smallFileList.iterator();
+      while (iterator.hasNext()) {
+        String smallFile = iterator.next();
+        if (smallFile == null || smallFile.isEmpty()
+            || smallFilesLock.contains(smallFile)) {
+          iterator.remove();
+        } else {
+          FileState fileState = new FileState(smallFile,
+              FileState.FileType.NORMAL, FileState.FileStage.PROCESSING);
+          try {
+            metaStore.insertUpdateFileState(fileState);
+          } catch (MetaStoreException e) {
+            LOG.error("Failed to insert file state.", e);
+          }
+        }
+      }
+
+      if (smallFileList.isEmpty()) {
+        return false;
+      }
+      smallFilesLock.addAll(smallFileList);
+
+      // Update container file map, save the info indicated whether it already exists
+      try {
+        if (dfsClient.exists(containerFilePath)) {
+          containerFileMap.put(
+              actionId, new ContainerFileExistInfo(containerFilePath, true));
```

Comment:
`ContainerFileExistInfo` denotes container file's status, it has no connection with action, actions can share the same info, so it should be indexed by action id.